qualtran/qir_example.ipynb

```diff
@@ -0,0 +1,685 @@
+    "    num_qubits = get_num_qubits_for_bloq(bloq)\n",
+    "    return Function(\n",
+    "    pyqir.FunctionType(void_type, [qubit_type]*num_qubits),\n",
+    "    Linkage.EXTERNAL,\n",
+    "    name,\n",
+    "    mod\n",
+    "    )\n",
+    "\n",
+    "def create_ir_map(bloq: qualtran.Bloq):\n",
+    "    param_counter = 0\n",
+    "    ir_map = {}\n",
+    "    # Loop through all registers in signature\n",
+    "    for register in bloq.signature.lefts():\n",
+    "        shape = register.shape[0] if len(register.shape) != 0 else 1 # get the shape as an int (we are asumming its 1d for simplicity)\n",
+    "        # map the (register_name, index_in_register) to the overall index\n",
+    "        ir_map.update({(register.name, i*register.bitsize + j): param_counter + i*register.bitsize + j for i in range(shape) for j in range(register.bitsize)})\n",
+    "        param_counter += register.bitsize * shape\n",
+    "    return ir_map\n",
+    "\n",
+    "\n",
+    "def find_ir_for_index(i, ir_map):\n",
+    "    for key, val in ir_map.items():\n",
+    "        if i == val:\n",
+    "            return key\n",
+    "    raise Exception(f\"Error in find_ir_for_index: No IR for index {i}\")\n",
+    "\n",
+    "def get_indexes_from_reg(ir_map, target_reg_name):\n",
+    "    return [ir_map[(reg_name, i)] for (reg_name, i) in ir_map.keys() if reg_name == target_reg_name]\n",
+    "\n",
+    "def generate_irs_from_soquet(soq):\n",
+    "    reg_name = soq.reg.name\n",
+    "    starting_index = soq.idx[0] if len(soq.idx) != 0 else 0\n",
+    "    return [(reg_name, starting_index*soq.reg.bitsize + i) for i in range(soq.reg.bitsize)]\n",
+    "\n",
+    "def map_soquet_to_params(soq, soq_map, ir_map):\n",
+    "    if isinstance(soq, np.ndarray):\n",
+    "        # renaming to soqs for clarity\n",
+    "        soqs = soq\n",
+    "        return [el for soq in soqs for el in map_soquet_to_params_helper(soq, soq_map, ir_map)]\n",
+    "    return map_soquet_to_params_helper(soq, soq_map, ir_map)\n",
+    "\n",
+    "def map_soquet_to_params_helper(soquet, soq_map, ir_map):\n",
+    "    if soquet in soq_map:\n",
+    "        irs = soq_map[soquet]\n",
+    "        return [ir_map[ir] for ir in irs]\n",
+    "    irs = generate_irs_from_soquet(soquet)\n",
+    "    return [ir_map[ir] for ir in irs]\n",
+    "\n",
+    "def compile_bloq(bloq: qualtran.Bloq, qubit_type, void_type, module, context, builder, func_dict=dict()):\n",
+    "    func_name = f\"{bloq.pretty_name()}_{get_num_qubits_for_bloq(bloq)}\"\n",
+    "    if func_name in func_dict:\n",
+    "        return func_dict[func_name]\n",
+    "    bloq_func = create_func_for_bloq(bloq, func_name, qubit_type, void_type, module)\n",
+    "    ir_map = create_ir_map(bloq)\n",
+    "    soq_map = {}\n",
+    "\n",
+    "    # It seems like QFTTextBook and PhaseGradientUnitary can be decomposed even though supports_decompose_bloq returns false\n",
```

Comment:
Seems like a bug in the `GateWithRegisters` base class. We should override this method for the `GateWithRegister` class to check whether the sub-bloq defines a cirq style decomposition and if yes, the return True from `supports_decompose_bloq`. 

Can you open a separate issue to track this?

cc @mpharrigan